cocotb/decorators.py

```diff
@@ -111,9 +111,10 @@ def __init__(self, inst, parent):
         self._outcome = None
 
         if not hasattr(self._coro, "send"):
-            self.log.error("%s isn't a valid coroutine! Did you use the yield "
-                           "keyword?" % self.funcname)
-            raise CoroutineComplete()
+            raise TypeError(
+                "%s isn't a valid coroutine! Did you use the yield "
+                "keyword?" % self.funcname
```

Comment:
Do async functions end up in the same code path? In this case, adding a `yield` would be the wrong suggestion. (Doesn't need to be handled in this PR.)

Reply:
This can never happen for async functions. Forgetting the `async` keyword would hit this path, but then that's not an async function!

Elaborating on that thought: this:
```
async def foo():
    return 1
```
is equivalent to 
```
@cocotb.coroutine
def foo():
    return 1
    yield
```

Reply:
Ah of course, thanks for clarifying that!